typespec/UCP/resourceproviders.tsp

```diff
@@ -0,0 +1,323 @@
+
+@doc("The resource type name. Example: 'redisCaches'.")
+@maxLength(63)
+@pattern("^([A-Za-z]([-A-Za-z0-9]*[A-Za-z0-9]))$")
+scalar ResourceTypeNameString extends string;
+
+@doc("The resource type API version. Example: '2023-10-01-preview'.")
+@maxLength(63)
+@pattern("^\\d{4}-\\d{2}-\\d{2}(-preview)?$")
+scalar ApiVersionNameString extends string;
+
+@doc("The resource provider location name. Example: 'eastus'.")
+@maxLength(63)
+@pattern("^[A-Za-z][A-Za-z0-9]*$")
+scalar LocationNameString extends string;
+
+#suppress "@azure-tools/typespec-azure-resource-manager/arm-resource-path-segment-invalid-chars"
+@doc("The resource type for defining a resource provider.")
+model ResourceProviderResource is TrackedResource<ResourceProviderProperties> {
+  @key("resourceProviderName")
+  @doc("The resource provider name. This is also the resource provider namespace. Example: 'Applications.Datastores'.")
+  @path
+  @segment("providers/System.Resources/resourceproviders")
+  name: ResourceProviderNamespaceString;
+}
+
+@doc("The properties of a resource provider.")
+model ResourceProviderProperties {
+  @doc("The status of the asynchronous operation.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+}
+
+@doc("The resource type for defining a resource type supported by the containing resource provider.")
+model ResourceTypeResource is ProxyResource<ResourceTypeProperties> {
+  @key("resourceTypeName")
+  @doc("The resource type name.")
+  @path
+  @segment("resourcetypes")
+  name: ResourceTypeNameString;
+}
+
+@doc("The properties of a resource type.")
+model ResourceTypeProperties {
+  @doc("The status of the asynchronous operation.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+
+  @doc("The default api version for the resource type.")
+  defaultApiVersion: ApiVersionNameString;
+}
+
+@doc("The resource type for defining an API version of a resource type supported by the containing resource provider.")
+model ApiVersionResource is ProxyResource<ApiVersionProperties> {
+  @key("apiVersionName")
+  @doc("The API version name.")
+  @path
+  @segment("apiversions")
+  name: ApiVersionNameString;
+}
+
+@doc("The properties of an API version.")
+model ApiVersionProperties {
+  @doc("The status of the asynchronous operation.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+}
+
+@doc("The resource type for defining a location of the containing resource provider. The location resource represents a logical location where the resource provider operates.")
+model LocationResource is ProxyResource<LocationProperties> {
+  @key("locationName")
+  @doc("The location name.")
+  @path
+  @segment("locations")
+  name: LocationNameString;
+}
+
+@doc("The properties of a location.")
+model LocationProperties {
+  @doc("The status of the asynchronous operation.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+
+  @doc("Address of a resource provider implementation.")
+  address?: string;
```

Comment:
Is it possible that this is not populated? should this be required field?

Reply:
This should be optional. It will be populated when there's a "full" RP implementation - like our existing RPs. It won't be populated when the RP is "virtual".